datafusion/optimizer/src/decorrelate_where_in.rs

```diff
@@ -121,80 +121,95 @@ impl OptimizerRule for DecorrelateWhereIn {
-        // split into filters
-        let subqry_filter_exprs = split_conjunction(&subqry_filter.predicate);
-        verify_not_disjunction(&subqry_filter_exprs)?;
-
-        // Grab column names to join on
-        let (col_exprs, other_exprs) =
-            find_join_exprs(subqry_filter_exprs, subqry_filter.input.schema())
-                .map_err(|e| context!("column correlation not found", e))?;
-        if !col_exprs.is_empty() {
-            // it's correlated
-            subqry_input = subqry_filter.input.clone();
-            (outer_cols, subqry_cols, join_filters) =
-                exprs_to_join_cols(&col_exprs, subqry_filter.input.schema(), false)
-                    .map_err(|e| context!("column correlation not found", e))?;
-            other_subqry_exprs = other_exprs;
-        }
-    }
 
-    let (subqry_cols, outer_cols) =
-        merge_cols((&[subquery_col], &subqry_cols), (&[outer_col], &outer_cols));
-
-    // build subquery side of join - the thing the subquery was querying
-    let subqry_alias = alias.next("__correlated_sq");
-    let mut subqry_plan = LogicalPlanBuilder::from((*subqry_input).clone());
-    if let Some(expr) = conjunction(other_subqry_exprs) {
-        // if the subquery had additional expressions, restore them
-        subqry_plan = subqry_plan.filter(expr)?
+    // extract join filters
+    let (join_filters, subquery_input) = extract_join_filters(subquery_input.as_ref())?;
+
+    // in_predicate may be also include in the join filters, remove it from the join filters.
+    let in_predicate = Expr::eq(query_info.where_in_expr.clone(), subquery_expr.clone());
+    let join_filters = remove_duplicate_filter(join_filters, in_predicate);
+
+    // replace qualified name with subquery alias.
+    let subquery_alias = alias.next("__correlated_sq");
+    let input_schema = subquery_input.schema();
+    let mut subquery_cols =
+        join_filters
+            .iter()
+            .try_fold(BTreeSet::<Column>::new(), |mut cols, expr| {
+                let using_cols: Vec<Column> = expr
+                    .to_columns()?
+                    .into_iter()
+                    .filter(|col| input_schema.field_from_column(col).is_ok())
+                    .collect::<_>();
+
+                cols.extend(using_cols);
+                Result::Ok(cols)
+            })?;
+    let join_filter = conjunction(join_filters).map_or(Ok(None), |filter| {
+        replace_qualify_name(filter, &subquery_cols, &subquery_alias).map(Option::Some)
+    })?;
+
+    // add projection
+    if let Expr::Column(col) = subquery_expr {
+        subquery_cols.remove(col);
     }
-    let projection = alias_cols(&subqry_cols);
-    let subqry_plan = subqry_plan
-        .project(projection)?
-        .alias(&subqry_alias)?
+    let subquery_expr_name = format!("{:?}", unnormalize_col(subquery_expr.clone()));
+    let first_expr = subquery_expr.clone().alias(subquery_expr_name.clone());
```

Comment:
For the right side of the in predicate, the original qualify name seems strange, so it is stripped.